lib/Differentiator/ReverseModeVisitor.cpp

```diff
@@ -133,30 +135,220 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
+              clad_compat::Expr_EvaluateAsConstantExpr(L, dummy, m_Context);
+          bool isConstR =
+              clad_compat::Expr_EvaluateAsConstantExpr(R, dummy, m_Context);
+
+          if (isConstL && isConstR)
+            return false;
+
+          if (!isConstL && isConstR)
+            return TraverseStmt(L);
+
+          if (isConstL && !isConstR)
+            return TraverseStmt(R);
+
+          return isInjectiveE(BinOp);
+        }
+        return false;
+      }
+
+      bool TraverseDeclRefExpr(clang::DeclRefExpr* DRE) {
+        if (auto* VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+          if (auto* init = VD->getInit())
+            return isInjectiveE(init->IgnoreImpCasts());
+        }
+        return false;
+      }
+
+      bool TraverseIntegerLiteral(clang::IntegerLiteral* IL) { return false; }
+
+    } checker(ctx);
+
+    return checker.isInjectiveIdx(E);
+  }
+
   bool ReverseModeVisitor::shouldUseCudaAtomicOps(const Expr* E) {
     if (!m_Context.getLangOpts().CUDA)
       return false;
-
-    if (!isa<DeclRefExpr>(E))
-      return false;
-
-    const auto* DRE = cast<DeclRefExpr>(E);
-
-    if (const auto* PVD = dyn_cast<ParmVarDecl>(DRE->getDecl())) {
-      if (m_DiffReq->hasAttr<clang::CUDAGlobalAttr>())
-        // Check whether this param is in the global memory of the GPU
-        return m_DiffReq.HasIndependentParameter(PVD);
-      if (m_DiffReq->hasAttr<clang::CUDADeviceAttr>()) {
-        for (auto index : m_DiffReq.CUDAGlobalArgsIndexes) {
-          const auto* PVDOrig = m_DiffReq->getParamDecl(index);
-          if ("_d_" + PVDOrig->getNameAsString() == PVD->getNameAsString() &&
-              (utils::isArrayOrPointerType(PVDOrig->getType()) ||
-               PVDOrig->getType()->isReferenceType()))
-            return true;
+    if (const auto* DRE = dyn_cast<DeclRefExpr>(E)) {
+      if (const auto* PVD = dyn_cast<ParmVarDecl>(DRE->getDecl())) {
+        if (m_DiffReq->hasAttr<clang::CUDAGlobalAttr>())
+          // Check whether this param is in the global memory of the GPU
+          return m_DiffReq.HasIndependentParameter(PVD);
+        if (m_DiffReq->hasAttr<clang::CUDADeviceAttr>()) {
+          for (auto index : m_DiffReq.CUDAGlobalArgsIndexes) {
+            const auto* PVDOrig = m_DiffReq->getParamDecl(index);
+            if ("_d_" + PVDOrig->getNameAsString() == PVD->getNameAsString() &&
+                (utils::isArrayOrPointerType(PVDOrig->getType()) ||
+                 PVDOrig->getType()->isReferenceType()))
+              return true;
+          }
+        }
+      }
+    } else if (const auto* ASE = dyn_cast<ArraySubscriptExpr>(E)) {
+      if (m_DiffReq->hasAttr<clang::CUDAGlobalAttr>()) {
+        const auto* idx = ASE->getIdx();
+        const auto* base =
+            dyn_cast<DeclRefExpr>(ASE->getBase()->IgnoreImpCasts());
+        if (const auto* PVD = dyn_cast<ParmVarDecl>(base->getDecl())) {
+          if (m_DiffReq->hasAttr<clang::CUDAGlobalAttr>()) {
```

Comment:
If we only check the const indexes of arrays, we can change that to check the CUDAGlobalArgsIndexes list so we can cover the device pullbacks as well.
```suggestion
          if (!m_DiffReq.CUDAGlobalArgsIndexes.empty()) {
```

Reply:
We have discussed this I think, apparently adding support for the device functions isn't that easy, so this should be a subject of sone othre PR

Reply:
For pullbacks as well. We need to make sure that the memory is global, which requires more then checking whether there are any elements in `m_DiffReq.CUDAGlobalArgsIndexes`.